File: src/tools/dispatch.ts
import type { AgentRequest, AgentResult, ToolName } from "./types.js";
import { runPi } from "./pi.js";
import { runCursor } from "./cursor.js";
import { runClaude } from "./claude.js";
import { runOpenCode } from "./opencode.js";
import { runAntigravity } from "./antigravity.js";
import { QcError } from "../errors.js";

/** Deep seam: one call site for every headless JSON adapter. */
export async function runAgent(tool: ToolName, request: AgentRequest): Promise<AgentResult> {
  switch (tool) {
    case "pi":
      return runPi(request);
    case "cursor":
      return runCursor(request);
    case "claude":
      return runClaude(request);
    case "opencode":
      return runOpenCode(request);
    case "antigravity":
      return runAntigravity(request);
    default: {
      const _exhaustive: never = tool;
      throw new QcError(`unknown tool '${String(_exhaustive)}'`);
    }
  }
}

File: src/tools/dispatch.rs
use crate::errors::QcError;
use crate::tools::antigravity::run_antigravity;
use crate::tools::claude::run_claude;
use crate::tools::codex::run_codex;
use crate::tools::cursor::run_cursor;
use crate::tools::opencode::run_opencode;
use crate::tools::pi::run_pi;
use crate::tools::types::{AgentRequest, AgentResult, ToolName};

/// Deep seam: one call site for every headless JSON adapter.
pub fn run_agent(tool: ToolName, request: &AgentRequest) -> Result<AgentResult, QcError> {
    match tool {
        ToolName::Pi => run_pi(request),
        ToolName::Cursor => run_cursor(request),
        ToolName::Claude => run_claude(request),
        ToolName::Opencode => run_opencode(request),
        ToolName::Antigravity => run_antigravity(request),
        ToolName::Codex => run_codex(request),
    }
}
